momepy/preprocessing.py

```diff
@@ -740,3 +745,368 @@ def _get_extrapolated_line(coords, tolerance, point=False):
+    )
+    edges_gdf[edge_to_col].where(
+        edges_gdf[edge_to_col] >= 0, edges_gdf["original_to"], inplace=True
+    )
+
+    # Get the edge list with (from, to, data):
+    new_edges_list = list(
+        zip(
+            edges_gdf[edge_from_col],
+            edges_gdf[edge_to_col],
+            edges_gdf.iloc[:, 2:].to_dict("index").values(),
+        )
+    )
+
+    return new_edges_list
+
+
+def _linear_simplification(geometry, new_origin, new_destination):
+    """
+    Rebuilds edge given (potential) new origin and destinations through a linear method: current line is linearly extended
+     to and from the new endpoints.
+
+    :param geometry: shapely.LineString, current edge
+    :param new_origin: shapely.Point or None, new origin point or None if origin remains the same
+    :param new_destination: shapely.Point or None, new destination point or None if destination remains the same
+
+    return shapely.LineString
+    """
+    # If we are dealing with a self-loop the line has no endpoints:
+    if new_origin == new_destination:
+        current_node = Point(line_coords[0])
+        geometry = linemerge([LineString([new_origin, current_node]), geometry])
+    # Assuming the line is not closed, we can find its endpoints:
+    else:
+        current_origin, current_destination = geometry.boundary
+        if new_origin is not None:
+            geometry = linemerge([LineString([new_origin, current_origin]), geometry])
+        if new_destination is not None:
+            geometry = linemerge(
+                [geometry, LineString([current_destination, new_destination])]
+            )
+    return geometry
+
+
+def _spider_simplification(geometry, new_origin, new_destination, buff=15):
+    """
+    Rebuilds edge giving (potential) new origin and destinations through a spide-web method: current line is broken within
+     small buffer from new endpoint and linearly extended.
+
+    :param geometry: shapely.LineString, current edge
+    :param new_origin: shapely.Point or None, new origin point or None if origin remains the same
+    :param new_destination: shapely.Point or None, new destination point or None if destination remains the same
+
+    return shapely.LineString
+    """
+    # If we are dealing with a self-loop the line has no boundary and we just use the first coordinate:
+    if new_origin == new_destination:
+        current_node = Point(line_coords[0])
+        geometry = linemerge([LineString([new_origin, current_node]), geometry])
+    # Assuming the line is not closed, we can find its endpoints via the boundary attribute:
+    else:
+        current_origin, current_destination = geometry.boundary
+        if new_origin is not None:
+            # Create a buffer around the new origin:
+            new_origin_buffer = new_origin.buffer(buff)
+            # Use shapely.ops.split to break the edge where it intersects the buffer:
+            geometry_split_by_buffer_list = list(split(geometry, new_origin_buffer))
```

Comment:
have you tried using `difference` instead? It may be more efficient that `split` + `linemerge` (but may not and may cause more trouble than good).